### PR TITLE
Give users the option to add symbols

### PR DESCRIPTION
**Pull request by technetium**
As suggested in issue #54

I've added a "+"-button at the right side of the search field. Feel free to style the button as you like.

All symbols can be edited by double clicking it. Double clicking the name will open a editbox for the name.
Clicking the trashcan will remove the symbol. (Trashcan is styled in CSS, feel free to change it)

Suggestion: Change the text: "Open a PR!" to "Add it yourself" with a link to the addSymbol function.
That will save to some pull requests and discussions about what symbols to include.
Then the only discussion will be what will be a good set of default symbols.

Also made the symbols draggable, so users can get them in their prefered order.
When a symbol is dragged the opacity is reduced to 40%, 
use the drag class to change the style of dragged symbols.
The target of the symbol gets a solid border,
the class over can be used to change that.

### CODE REVIEW

File: symbols.js
Comment (samwho) [inline symbols.js:416]: Could we make it mandatory that a symbol have a code point at the 0th position? As in, if they hit enter with an empty text box the symbol either reverts to its default state (for existing symbols) or disappears entirely (for new symbols)?

File: symbols.js
Comment (samwho) [inline symbols.js:405]:
```suggestion

```

I think this line is superfluous, `symbols` is set to the empty array on 398 already.

File: symbols.js
Comment (samwho) [inline symbols.js:441]: Indentation of this function is a little off, and this `console.log` should be removed before merging.

File: symbols.js
Comment (samwho) [inline symbols.js:624]: 👀

File: symbols.js
Comment (samwho) [inline symbols.js:619]: We could reduce a fair bit of density / visual noise here by introducing some variables:

```suggestion
if (target.classList?.contains("symbol")) {
  const parent = target.parentElement;
  const dataset = parent.dataset;

  dataset.dragTarget = Array.from(parent.children).indexOf(target);

  symbols.splice(dataset.dragTarget, 0, symbols.splice(dataset.dragIndex, 1)[0]);
  window.localStorage.setItem("symbols", JSON.stringify(symbols));

  if (parseInt(dataset.dragIndex) < parseInt(dataset.dragTarget)) {
    if (target.nextElementSibling) {
      target.nextElementSibling.before(parent.children[dataset.dragIndex]);
    } else {
      parent.appendChild(parent.children[dataset.dragIndex]);
    }
  } else {
    target.before(parent.children[dataset.dragIndex]);
  }
  break;
}
```

File: symbols.js
Comment (samwho) [inline symbols.js:655]:
```suggestion

```

File: symbols.js
Comment (samwho) [inline symbols.js:678]: Can likely remove this, the handler for it is empty.

File: symbols.css
Comment (samwho) [inline symbols.css:181]: In situ it's not immediately obvious to me that this is a trash can, could we try ❌ instead?

**Review by samwho — COMMENTED** [review]
This is a really good start! I've left a smattering of review comments.

I'm on the fence about editing the default symbols. On the one hand, it's unobtrusive. On the other I feel like it's fairly easy to screw things up in a way that's tricky to correct should you want to.

File: symbols.js
Comment (technetium) [reply symbols.js:416]: New symbols are created with an empty code point. I can change that to a space (or any other symbol), but then users have to delete the space when editing a new symbol.
Other option is to use a magic symbol, an remove that symbol before the user start editing.  Disadvantages: the magic symbol cannot be used, and extra code needed to check for the magic character.
Adding the question mark is the simplest solution.

File: symbols.js
Comment (technetium) [reply symbols.js]: OK, implemented in 914415c9dd6387c7983bcf6ddd23508b77b5c2fe

File: symbols.js
Comment (technetium) [reply symbols.js:481]: This (and other) console.log removed in 914415c9dd6387c7983bcf6ddd23508b77b5c2fe

File: symbols.js
Comment (technetium) [reply symbols.js]: Gone in 914415c9dd6387c7983bcf6ddd23508b77b5c2fe ;-)

File: symbols.js
Comment (technetium) [reply symbols.js]: Added 3 extra variables in 914415c9dd6387c7983bcf6ddd23508b77b5c2fe
Code noise is even more reduces than in the suggestion.

File: symbols.js
Comment (technetium) [reply symbols.js]: dragover needs to be handled, else the elements are not droppable
dragleave can be removed and is in 914415c9dd6387c7983bcf6ddd23508b77b5c2fe
other handlers are not empty, they stay.

File: symbols.css
Comment (technetium) [reply symbols.css]: Whatever you want. Also in 914415c9dd6387c7983bcf6ddd23508b77b5c2fe

**Comment by technetium** [comment]
Preventing the deletion/change of the default symbols is extra code.
To restore the defaults is a matter of removing the symbols key from the local storage.
I'm on the fence regarding a button to reset to the default values:
A custom layout might be gone too easy.
Other option is to have the option of importing and exporting the symbols.
The website gets further and further away from being dumb.

File: symbols.js
Comment (samwho) [inline symbols.js:449]: I was thinking about this earlier today: what happens when the default symbols are updated? I think currently the user won't see them, they'll load what's in local storage and completely ignore the defaults.

We probably need some sort of merging. Alternately, we track users' custom symbols separately to the defaults.

**Review by samwho — CHANGES_REQUESTED** [review]

File: symbols.js
Comment (technetium) [reply symbols.js:462]: If you want to merge new defaults, with the existing symbols, we need to label either the user symbols or the defaults (which adds complexity) and then disable the editing of the defaults (which adds complexity) then it will be impossible for users to move the frequently used default symbols to the top of the list, resulting in users that will add existing default symbols also to the user symbols. Reducing the benefits of a list of default symbols.

Another method is to track all the changes made by the user (which adds way more complexity)  and apply them every time the defaults are loaded (which is so complex I cannot imagine how to do that)

If you can program that, you're welcome to do that. I probably don't have the knowledge, and if I did my time is also limited.

Another solution would be to publish sets of symbols, including a default. When a set (e.g. the default set) is updated let the user load that set into their collection. That's something I could implement in a few hours. As a matter of fact, I already did. It's in pull request #59

**Comment by samwho** [comment]
Sorry I've not gotten around to taking a closer look at this yet. I'll be travelling for work next week as well, so it's likely I won't get to this for a couple of weeks. I really appreciate the time you've put in to this, so I'm sorry for the delay.

**Comment by samwho** [comment]
@technetium sorry for the delay on this, but I've decided to not merge this PR. It adds a level of complexity I don't think I want to maintain in the tool. You have my blessing to fork and re-host if this is something that makes a substantial difference to your workflow. I appreciate the time and effort you've put in to contributing to symbol.wtf <3

**Pull request closed without merging**
